File: Packages/TriathlonEngine/Sources/TriathlonEngine/Predictions/Riegel.swift
import Foundation

/// Modèle de Riegel pour extrapoler un temps entre deux distances :
/// `T2 = T1 · (D2/D1)^b`, avec b ≈ 1.06 pour la course d'endurance.
public enum Riegel {
    public static let defaultExponent = 1.06

    public static func predictTimeSeconds(
        knownDistanceMeters d1: Double,
        knownTimeSeconds t1: Double,
        targetDistanceMeters d2: Double,
        exponent b: Double = defaultExponent
    ) -> Double {
        precondition(d1 > 0 && t1 > 0 && d2 > 0)
        return t1 * pow(d2 / d1, b)
    }

    /// Estime l'exposant de fatigue propre à l'athlète à partir de deux perfs.
    public static func fittedExponent(
        d1: Double, t1: Double, d2: Double, t2: Double
    ) -> Double {
        precondition(d1 > 0 && t1 > 0 && d2 > 0 && t2 > 0 && d1 != d2)
        return log(t2 / t1) / log(d2 / d1)
    }
}

File: web/src/engine/predictions/riegel.ts
// Portage de Predictions/Riegel.swift — T2 = T1·(D2/D1)^b, b ≈ 1.06.

export const RIEGEL_DEFAULT_EXPONENT = 1.06;

export function predictTimeSeconds(
  knownDistanceMeters: number,
  knownTimeSeconds: number,
  targetDistanceMeters: number,
  exponent: number = RIEGEL_DEFAULT_EXPONENT,
): number {
  if (!(knownDistanceMeters > 0 && knownTimeSeconds > 0 && targetDistanceMeters > 0)) {
    throw new Error('Riegel: distances & temps > 0');
  }
  return knownTimeSeconds * Math.pow(targetDistanceMeters / knownDistanceMeters, exponent);
}

/** Estime l'exposant de fatigue propre à l'athlète à partir de deux perfs. */
export function fittedExponent(d1: number, t1: number, d2: number, t2: number): number {
  if (!(d1 > 0 && t1 > 0 && d2 > 0 && t2 > 0 && d1 !== d2)) throw new Error('Riegel.fittedExponent: entrées invalides');
  return Math.log(t2 / t1) / Math.log(d2 / d1);
}

export const Riegel = { predictTimeSeconds, fittedExponent, defaultExponent: RIEGEL_DEFAULT_EXPONENT };
